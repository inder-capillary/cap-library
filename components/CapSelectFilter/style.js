import styled from 'styled-components';
import CapMenu from '../CapMenu';

import {
  CAP_SPACE_04,
  CAP_SPACE_08,
  CAP_SPACE_16,
  CAP_SPACE_32,
  CAP_G13,
  CAP_G01,
  CAP_RED,
  CAP_WHITE,
  FONT_COLOR_05,
  CAP_PALE_GREY,
  FONT_SIZE_S,
  FONT_SIZE_L,
  FONT_WEIGHT_MEDIUM,
} from '../styled/variables';

export const StyledDiv = styled.div`
    display: flex;
    align-items: center;
    padding: ${CAP_SPACE_08} ${CAP_SPACE_08} ${CAP_SPACE_08} ${CAP_SPACE_16};
    border: ${(props) => (props?.selected ? 'none' : `1px solid #9e9d87`)};
    border-radius: ${CAP_SPACE_16};
    height: ${CAP_SPACE_32};
    width: ${(props) => (props?.showBadge ? `calc(${props?.width} + 25px)` : props?.width)};
    background-color: ${(props) => props?.selected && CAP_G13};
    cursor: pointer;
    &.ant-dropdown-open {
        border: 1px solid ${CAP_G01};
    }
`;

export const MenuItem = styled(CapMenu.Item)`
    background-color: ${(props) => props?.selected && CAP_PALE_GREY};
    border-left: ${(props) => props?.selected && `3px solid ${FONT_COLOR_05}`};
`;

export const Badge = styled.div`
  display: flex;
  align-items: center;
  background-color: ${CAP_RED};
  color: ${CAP_WHITE};
  height: ${FONT_SIZE_L};
  width: ${FONT_SIZE_L};
  border-radius: 50%;
  margin: 0 ${CAP_SPACE_04};
  padding: 0 5px;
  font-size: ${FONT_SIZE_S};
  font-weight: ${FONT_WEIGHT_MEDIUM};
  line-height: normal;
`;
