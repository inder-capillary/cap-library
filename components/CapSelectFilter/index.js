/**
*
* CapSelectFilter
*
*/

import React from 'react';
import PropTypes from "prop-types";
import classNames from 'classnames';
import CapDropdown from '../CapDropdown';
import CapLabel from '../CapLabel';
import CapMenu from '../CapMenu';
import CapHeading from '../CapHeading';
import CapIcon from '../CapIcon';
import { StyledDiv, MenuItem, Badge } from './style';
import {
  CAP_G12,
} from '../styled/variables';

const { CapLabelInline } = CapLabel;
const clsPrefix = 'cap-select-filter-v2';

function CapSelectFilter(props) {
  const {
    className,
    overlayClassName,
    onSelect,
    selectedValue,
    placeholder,
    placement,
    width,
    dropdownMaxHeight,
    dropdownWidth,
    data,
    showBadge = false,
    ...rest
  } = props;

  const getMenuItems = (dataArray = []) => {
    const menuItems = (
      <CapMenu style={{ border: `1px solid ${CAP_G12}` }}>
        {dataArray?.map(({ key, value, label }) => (
          <MenuItem
            key={key}
            value={value}
            onClick={onSelect}
            selected={value === selectedValue}
          >
            <CapHeading type="h5">{label}</CapHeading>
          </MenuItem>
        ))
        }
      </CapMenu>
    );
    return menuItems;
  };

  return (
    <CapDropdown
      placement={placement}
      trigger={['click']}
      overlay={getMenuItems(data)}
      className={classNames(clsPrefix, className)}
      overlayClassName={classNames(`${clsPrefix}-overlay`, overlayClassName)}
      overlayStyle={{
        maxHeight: dropdownMaxHeight,
        width: dropdownWidth,
        overflowY: 'scroll',
      }}
      getPopupContainer={(trigger) => trigger?.parentNode || document.body}
      {...rest}
    >
      <StyledDiv selected={selectedValue?.length} width={width} showBadge={showBadge && selectedValue}>
        <CapLabelInline type="label9">
          {placeholder}
        </CapLabelInline>
        {(showBadge && selectedValue) && (
          <Badge>{[selectedValue].length}</Badge>
        )}
        <CapIcon type="chevron-down" />
      </StyledDiv>
    </CapDropdown>
  );
}

CapSelectFilter.defaultProps = {
  placement: "bottomLeft",
  width: '80px',
  dropdownMaxHeight: '320px',
  dropdownWidth: '228px',
};

CapSelectFilter.propTypes = {
  placement: PropTypes.string,
  className: PropTypes.string,
  overlayClassName: PropTypes.string,
  onSelect: PropTypes.func,
  selectedValue: PropTypes.string,
  data: PropTypes.array,
  placeholder: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  width: PropTypes.string,
  dropdownMaxHeight: PropTypes.string,
  dropdownWidth: PropTypes.string,
};

export default CapSelectFilter;
