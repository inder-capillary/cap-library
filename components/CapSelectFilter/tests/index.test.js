
import React from 'react';
import { shallow } from 'enzyme';
import CapSelectFilter from '../index';

describe('<CapSelectFilter />', () => {
  let wrapper;
  const onSelect = jest.fn();
  const data = [
    {
      key: 'approved',
      value: 'APPROVED',
    },
    {
      key: 'completed',
      value: 'COMPLETED',
    },
    {
      key: 'disabled',
      value: 'DISABLED',
    }];

  const componentProps = {
    data,
    onSelect,
  };

  const renderFunction = (props) => shallow(<CapSelectFilter {...props} />);


  it('should render default component', () => {
    wrapper = renderFunction();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render default component with data', () => {
    wrapper = renderFunction(componentProps);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component selected value and badge', () => {
    componentProps.showBadge = true;
    componentProps.selectedValue = 'LIVE';
    wrapper = renderFunction(componentProps);
    expect(wrapper).toMatchSnapshot();
  });
});
