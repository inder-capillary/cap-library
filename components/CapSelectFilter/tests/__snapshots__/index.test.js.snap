// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CapSelectFilter /> should render component selected value and badge 1`] = `
<CapDropdown
  className="cap-select-filter-v2"
  getPopupContainer={[Function]}
  overlay={
    <CapMenu
      style={
        Object {
          "border": "1px solid #e8e8e8",
        }
      }
    >
      <ForwardRef(Styled(withItemHOC))
        onClick={[MockFunction]}
        selected={false}
        value="APPROVED"
      >
        <ForwardRef(styled.div)
          type="h5"
        >
          APPROVED
        </ForwardRef(styled.div)>
      </ForwardRef(Styled(withItemHOC))>
      <ForwardRef(Styled(withItemHOC))
        onClick={[MockFunction]}
        selected={false}
        value="COMPLETED"
      >
        <ForwardRef(styled.div)
          type="h5"
        >
          COMPLETED
        </ForwardRef(styled.div)>
      </ForwardRef(Styled(withItemHOC))>
      <ForwardRef(Styled(withItemHOC))
        onClick={[MockFunction]}
        selected={false}
        value="DISABLED"
      >
        <ForwardRef(styled.div)
          type="h5"
        >
          DISABLED
        </ForwardRef(styled.div)>
      </ForwardRef(Styled(withItemHOC))>
    </CapMenu>
  }
  overlayClassName="cap-select-filter-v2-overlay"
  overlayStyle={
    Object {
      "maxHeight": "320px",
      "overflowY": "scroll",
      "width": "228px",
    }
  }
  placement="bottomLeft"
  trigger={
    Array [
      "click",
    ]
  }
>
  <styled.div
    selected={4}
    showBadge="LIVE"
    width="80px"
  >
    <styled.span
      type="label9"
    />
    <styled.div>
      1
    </styled.div>
    <CapIcon
      type="chevron-down"
    />
  </styled.div>
</CapDropdown>
`;

exports[`<CapSelectFilter /> should render default component 1`] = `
<CapDropdown
  className="cap-select-filter-v2"
  getPopupContainer={[Function]}
  overlay={
    <CapMenu
      style={
        Object {
          "border": "1px solid #e8e8e8",
        }
      }
    />
  }
  overlayClassName="cap-select-filter-v2-overlay"
  overlayStyle={
    Object {
      "maxHeight": "320px",
      "overflowY": "scroll",
      "width": "228px",
    }
  }
  placement="bottomLeft"
  trigger={
    Array [
      "click",
    ]
  }
>
  <styled.div
    showBadge={false}
    width="80px"
  >
    <styled.span
      type="label9"
    />
    <CapIcon
      type="chevron-down"
    />
  </styled.div>
</CapDropdown>
`;

exports[`<CapSelectFilter /> should render default component with data 1`] = `
<CapDropdown
  className="cap-select-filter-v2"
  getPopupContainer={[Function]}
  overlay={
    <CapMenu
      style={
        Object {
          "border": "1px solid #e8e8e8",
        }
      }
    >
      <ForwardRef(Styled(withItemHOC))
        onClick={[MockFunction]}
        selected={false}
        value="APPROVED"
      >
        <ForwardRef(styled.div)
          type="h5"
        >
          APPROVED
        </ForwardRef(styled.div)>
      </ForwardRef(Styled(withItemHOC))>
      <ForwardRef(Styled(withItemHOC))
        onClick={[MockFunction]}
        selected={false}
        value="COMPLETED"
      >
        <ForwardRef(styled.div)
          type="h5"
        >
          COMPLETED
        </ForwardRef(styled.div)>
      </ForwardRef(Styled(withItemHOC))>
      <ForwardRef(Styled(withItemHOC))
        onClick={[MockFunction]}
        selected={false}
        value="DISABLED"
      >
        <ForwardRef(styled.div)
          type="h5"
        >
          DISABLED
        </ForwardRef(styled.div)>
      </ForwardRef(Styled(withItemHOC))>
    </CapMenu>
  }
  overlayClassName="cap-select-filter-v2-overlay"
  overlayStyle={
    Object {
      "maxHeight": "320px",
      "overflowY": "scroll",
      "width": "228px",
    }
  }
  placement="bottomLeft"
  trigger={
    Array [
      "click",
    ]
  }
>
  <styled.div
    showBadge={false}
    width="80px"
  >
    <styled.span
      type="label9"
    />
    <CapIcon
      type="chevron-down"
    />
  </styled.div>
</CapDropdown>
`;
